src/staking/KerosineDenominatorV3.sol

```diff
@@ -0,0 +1,148 @@
+    using FixedPointMathLib for uint256;
+
+    Kerosine public immutable KEROSINE;
+    Dyad public immutable DYAD;
+    KerosineManager public immutable MANAGER;
+
+    uint64 public dyadMultiplierSnapshot = 1e12;
+    uint64 public targetDyadMultiplier = 1e12;
+
+    uint32 public dyadMultiplierSnapshotTimestamp;
+    uint32 public targetDyadMultiplierTimestamp;
+
+    EnumerableSet.AddressSet private _excludedAddresses;
+
+    event DyadMultiplierUpdated(uint64 previous, uint64 target, uint32 fromTimestamp, uint32 toTimestamp);
+
+    error TargetMultiplierTooSmall();
+
+    constructor(Kerosine _kerosine, Dyad _dyad, KerosineManager _manager)
+        Owned(0xDeD796De6a14E255487191963dEe436c45995813)
+    {
+        KEROSINE = _kerosine;
+        DYAD = _dyad;
+        MANAGER = _manager;
+        _excludedAddresses.add(0xDeD796De6a14E255487191963dEe436c45995813); // Team Multisig
+        _excludedAddresses.add(0x3962f6585946823440d274aD7C719B02b49DE51E); // Sablier Linear Lockup
+    }
+
+    function setAddressExcluded(address _address, bool exclude) external onlyOwner {
+        if (exclude) {
+            _excludedAddresses.add(_address);
+        } else {
+            _excludedAddresses.remove(_address);
+        }
+    }
+
+    function setTargetDyadMultiplier(uint64 _targetMultiplier, uint32 _duration) external onlyOwner {
+        if (_targetMultiplier < 1e12) {
+            revert TargetMultiplierTooSmall();
+        }
+
+        uint64 previousMultiplier = _getDyadSupplyMultiplier();
+
+        dyadMultiplierSnapshot = previousMultiplier;
+        targetDyadMultiplier = _targetMultiplier;
+        dyadMultiplierSnapshotTimestamp = uint32(block.timestamp);
+        targetDyadMultiplierTimestamp = uint32(block.timestamp) + _duration;
+
+        emit DyadMultiplierUpdated(
+            previousMultiplier, _targetMultiplier, uint32(block.timestamp), uint32(block.timestamp) + _duration
+        );
+    }
+
+    function currentDyadMultiplier() external view returns (uint64) {
+        return _getDyadSupplyMultiplier();
+    }
+
+    function isExcludedAddress(address _address) external view returns (bool) {
+        return _excludedAddresses.contains(_address);
+    }
+
+    function excludedAddresses() external view returns (address[] memory) {
+        return _excludedAddresses.values();
+    }
+
+    function denominator() external view returns (uint256) {
+        uint256 adjustedKerosineSupply = KEROSINE.totalSupply();
+        uint256 excludedAddressLength = _excludedAddresses.length();
+        for (uint256 i = 0; i < excludedAddressLength; ++i) {
+            adjustedKerosineSupply -= KEROSINE.balanceOf(_excludedAddresses.at(i));
+        }
+
+        uint256 dyadMultiplier = _getDyadSupplyMultiplier();
+
+        // Return early if multiplier is 1
+        if (dyadMultiplier == 1e12) {
+            return adjustedKerosineSupply;
+        }
+
+        uint256 dyadSupply = DYAD.totalSupply();
+
+        uint256 tvl = _getTvl();
```

Comment:
This is a super expensive operation - are we able to bypass the check for TVL and instead just multiply the adjusted kerosene supply by dyadMultiplier?

For instance for 1.25x:

`x = (c - d) / k`
becomes
`x = (c - d) / (1.25 * k)`

It seems to me this should have the same effect?

Reply:
The original idea for the new kerosine deterministic value was:

`x = (c - 1.25 * d) / k`

Since the `KeroseneVault` contract is not upgradeable we cannot modify directly how the kero DV is calculated there, the only way was to use a new `KerosineDenominator` contract and calculate the denominator in such way that

`(c - d) / new denom = (c - 1.25*d) / k`